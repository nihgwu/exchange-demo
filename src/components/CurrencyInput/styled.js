import styled from 'styled-components';

export const Input = styled.input`
  background-color: transparent;
  color: ${props => props.theme.colors.white};
  border: none;
  outline: none;
  text-align: right;
`;
