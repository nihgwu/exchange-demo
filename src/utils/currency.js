export const currencies = {
  GBP: {
    code: 'GBP',
    symbol: '£',
    description: 'British Pound',
  },
  EUR: {
    code: 'EUR',
    symbol: '€',
    description: 'Euro',
  },
  USD: {
    code: 'USD',
    symbol: '$',
    description: 'American Dollar',
  },
};

export const currencyCodes = Object.keys(currencies);

export const getCodeIndex = code => currencyCodes.indexOf(code);

export const getNextCode = code => {
  const idx = getCodeIndex(code);
  let nextIdx = idx + 1;
  if (nextIdx >= currencyCodes.length) nextIdx = 0;

  return currencyCodes[nextIdx];
};

export const getCodeSymbol = code => {
  const currency = currencies[code];
  return currency ? currency.symbol : '';
};

// trim trailing zeros
export const trimNumber = (value, maxDigits = 2) => {
  return Number(value.toFixed(maxDigits));
};

export const formatMoney = (code, value) => {
  return `${getCodeSymbol(code)}${trimNumber(value)}`;
};

export const formatRate = (base, code, rates) => {
  if (!rates || !rates[code] || !rates[code]) return '';

  const rate = rates[code] / rates[base];
  const value = trimNumber(rate, 4);
  return `${getCodeSymbol(base)}1 = ${getCodeSymbol(code)}${value}`;
};
