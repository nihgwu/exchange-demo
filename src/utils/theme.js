import { createGlobalStyle } from 'styled-components';
import { normalize } from 'styled-normalize';

export const variant = (prop, variants) => props => variants[props[prop]];
export const boolean = (prop, style) => props => props[prop] && style;

export const GlobalStyle = createGlobalStyle`
  ${normalize};

  *,
  *:before,
  *:after {
    box-sizing: border-box;
  }
  
  body {
    margin: 0;
    padding: 0;
    font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", "Roboto", "Oxygen",
      "Ubuntu", "Cantarell", "Fira Sans", "Droid Sans", "Helvetica Neue",
      sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }

  .slider {
    .paging-dot circle {
      fill: white !important;
    }

    .slider-list {
      margin: 0 !important;
    }
  }
`;

export const colors = {
  primary: '#3b7afa',
  primaryActive: '#1c4eb3',
  white: '#fff',
  light: 'rgba(255,255,255,0.6)',
};

export default {
  colors,
};
