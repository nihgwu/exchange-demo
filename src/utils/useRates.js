import { useRef, useState, useEffect } from 'react';

import { getRates, cancelGetRates } from 'services/rate';

const TIMEOUT = 10 * 1000;

const useRates = () => {
  const initial = useRef(true);
  const [data, setData] = useState({
    loading: true,
    rates: null,
    error: null,
  });

  useEffect(() => {
    let interval;
    if (initial.current) {
      const callback = async () => {
        // cancel pending request
        cancelGetRates();
        try {
          const rates = await getRates();
          // TODO: we could avoid triggering re-render if rates stays the same
          setData({
            loading: false,
            rates,
            error: null,
          });
        } catch (error) {
          setData({
            loading: false,
            rates: null,
            error,
          });

          // TODO: if failed to fetch rates, retry immediately
          // clearInterval(interval);
          // interval = setInterval(callback, TIMEOUT);
          // callback();
        }
      };

      interval = setInterval(callback, TIMEOUT);
      callback();
      initial.current = false;
    }

    return () => clearInterval(interval);
  }, [setData]);

  return data;
};

export default useRates;
