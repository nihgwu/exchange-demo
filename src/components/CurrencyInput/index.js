import React, {
  forwardRef,
  useRef,
  useState,
  useCallback,
  useEffect,
} from 'react';
import PropTypes from 'prop-types';

import { trimNumber } from 'utils/currency';

import { Input } from './styled';

const isValidInput = input => input === '' || /^\d+\.?\d{0,2}$/.test(input);

const normalizeValue = value => {
  if (!value) return '';

  return trimNumber(value).toString();
}

const CurrencyInput = forwardRef(
  ({ value: _value, maxValue, prefix = '', onChange, ...rest }, ref) => {
    const prevValue = useRef(_value);
    const [value, setValue] = useState(normalizeValue(_value));
    const handleChange = useCallback(
      e => {
        const input = e.target.value.replace(prefix, '');

        if (isValidInput(input)) {
          const value = Number(input);
          if (maxValue !== undefined && value > maxValue) return;

          setValue(input);
          onChange && onChange(Number(value));
        }
      },
      [prefix, maxValue, setValue, onChange]
    );

    const handleBlur = useCallback(() => {
      if (Number(value) === 0) {
        setValue('');
      } else if (value[value.length - 1] === '.') {
        setValue(value.substr(0, value.length - 1));
      }
    }, [value]);

    useEffect(() => {
      if (prevValue.current !== _value) {
        setValue(normalizeValue(_value));
        prevValue.current = _value;
      }
    }, [prevValue, _value]);

    return (
      <Input
        {...rest}
        ref={ref}
        value={value === '' ? '' : `${prefix}${value}`}
        onChange={handleChange}
        onBlur={handleBlur}
      />
    );
  }
);

CurrencyInput.displayName = 'CurrencyInput';
CurrencyInput.propTypes = {
  value: PropTypes.number,
  maxValue: PropTypes.number,
  prefix: PropTypes.string,
  onChange: PropTypes.func,
};

export default CurrencyInput;
