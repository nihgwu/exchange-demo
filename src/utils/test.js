import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import theme from 'utils/theme';

const ThemeWrapper = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui, options) =>
  render(ui, { wrapper: ThemeWrapper, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
