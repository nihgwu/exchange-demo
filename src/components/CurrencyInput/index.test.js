import React from 'react';
import { render, fireEvent } from 'utils/test';

import CurrencyInput from './index';

describe('CurrencyInput', () => {
  it('should renders correctly', () => {
    const { container } = render(<CurrencyInput />);
    const input = container.firstChild;

    expect(input).toMatchSnapshot();
  });

  it('should normalizeValue correctly', () => {
    const { container, rerender } = render(<CurrencyInput />);
    const input = container.firstChild;

    expect(input).toHaveValue('');

    rerender(<CurrencyInput value={0} />);
    expect(input).toHaveValue('');

    rerender(<CurrencyInput value={0.1111} />);
    expect(input).toHaveValue('0.11');
  });

  it('should prefix if value is not 0', () => {
    const { container, rerender } = render(<CurrencyInput prefix="$" />);
    const input = container.firstChild;

    expect(input).toHaveValue('');

    rerender(<CurrencyInput prefix="$" value={0} />);
    expect(input).toHaveValue('');

    rerender(<CurrencyInput prefix="$" value={0.1} />);
    expect(input).toHaveValue('$0.1');
  });

  it('should clear input if value is 0 on blur', () => {
    const { container } = render(<CurrencyInput prefix="$" />);
    const input = container.firstChild;

    fireEvent.change(input, { target: { value: '0.0' } });
    expect(input).toHaveValue('$0.0');

    fireEvent.blur(input);
    expect(input).toHaveValue('');
  });

  it('should remove trailing dot', () => {
    const { container } = render(<CurrencyInput prefix="$" />);
    const input = container.firstChild;

    fireEvent.change(input, { target: { value: '10.' } });
    expect(input).toHaveValue('$10.');

    fireEvent.blur(input);
    expect(input).toHaveValue('$10');
  });

  it('should only accept valid number input', () => {
    const handleChange = jest.fn();
    const { container } = render(<CurrencyInput prefix="$" onChange={handleChange} />);
    const input = container.firstChild;

    fireEvent.change(input, { target: { value: '10' } });
    expect(input).toHaveValue('$10');
    expect(handleChange).toHaveBeenCalledWith(10);

    fireEvent.change(input, { target: { value: '10.$' } });
    expect(input).toHaveValue('$10.');
    expect(handleChange).toHaveBeenCalledWith(10);

    expect(handleChange).toHaveBeenCalledTimes(2);
    fireEvent.change(input, { target: { value: '10..' } });
    expect(input).toHaveValue('$10.');
    expect(handleChange).toHaveBeenCalledTimes(2);

    fireEvent.change(input, { target: { value: '10.1' } });
    expect(input).toHaveValue('$10.1');
    expect(handleChange).toHaveBeenCalledWith(10.1);

    expect(handleChange).toHaveBeenCalledTimes(3);
    fireEvent.change(input, { target: { value: '10.1.' } });
    expect(input).toHaveValue('$10.1');
    expect(handleChange).toHaveBeenCalledTimes(3);

    fireEvent.change(input, { target: { value: '10.11' } });
    expect(input).toHaveValue('$10.11');
    expect(handleChange).toHaveBeenCalledWith(10.11);

    expect(handleChange).toHaveBeenCalledTimes(4);
    fireEvent.change(input, { target: { value: '10.111' } });
    expect(input).toHaveValue('$10.11');
    expect(handleChange).toHaveBeenCalledTimes(4);
  });
});
