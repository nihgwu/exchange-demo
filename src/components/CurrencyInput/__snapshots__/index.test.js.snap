// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput should renders correctly 1`] = `
.c0 {
  background-color: transparent;
  color: #fff;
  border: none;
  outline: none;
  text-align: right;
}

<input
  class="c0"
  value=""
/>
`;
